source/extensions/filters/http/grpc_json_transcoder/config.cc

```diff
@@ -28,8 +29,49 @@ GrpcJsonTranscoderFilterConfig::createRouteSpecificFilterConfigTyped(
     const envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder&
         proto_config,
     Server::Configuration::ServerFactoryContext& context, ProtobufMessage::ValidationVisitor&) {
+  return createConfig(proto_config, context);
+}
+
+JsonTranscoderConfigSharedPtr GrpcJsonTranscoderFilterConfig::createConfig(
+    const envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder&
+        proto_config,
+    Server::Configuration::CommonFactoryContext& context) {
+  // Construct descriptor_pool_builder differently depending on whether we need
+  // reflection or not.
+  DescriptorPoolBuilderSharedPtr descriptor_pool_builder;
+  if (proto_config.descriptor_set_case() ==
+      envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
+          DescriptorSetCase::kReflectionClusterConfig) {
+
+    // Create and inject as many async clients as there are services, since we
+    // need one per service for parallelism, per qiwzhang's suggestion:
+    // https://github.com/envoyproxy/envoy/issues/1182#issuecomment-1026048698
+    envoy::config::core::v3::GrpcService config;
+    config.mutable_envoy_grpc()->set_cluster_name(
+        proto_config.reflection_cluster_config().cluster_name());
+    std::vector<Envoy::Grpc::RawAsyncClientSharedPtr> async_clients;
+    async_clients.reserve(proto_config.services_size());
+
+    for (int i = 0; i < proto_config.services_size(); i++) {
+      async_clients.push_back(std::make_shared<Envoy::Grpc::AsyncClientImpl>(
+          context.clusterManager(), config, context.api().timeSource()));
+    }
+
+    AsyncReflectionFetcherSharedPtr async_reflection_fetcher =
+        std::make_shared<AsyncReflectionFetcher>(proto_config.reflection_cluster_config(),
+                                                 proto_config.services(), async_clients,
```

Comment:
async_clients should be implementation details of AsyncReflectionFetcher.  It should be inside the class